dev/source-release.sh

```diff
@@ -66,7 +66,12 @@ tarball=$tag.tar.gz
 
 # be conservative and use the release hash, even though git produces the same
 # archive (identical hashes) using the scm tag
-git archive $release_hash --prefix $tag/ -o $tarball .baseline api arrow bundled-guava common core data dev flink gradle gradlew hive mr orc parquet pig spark spark2 spark-runtime spark3 spark3-runtime LICENSE NOTICE README.md build.gradle baseline.gradle deploy.gradle tasks.gradle jmh.gradle gradle.properties settings.gradle versions.lock versions.props version.txt
+adds=" .baseline"  # prefixed with a blank space for each file name including the first one
+excludes="build|examples|jitpack.yml|python|site"  # excluded as they are not of use for releasing jars
+echo "Excluded files and directories: ${excludes}"
+archives=$(git ls-tree --name-only -r HEAD | cut -d"/" -f1 | uniq | grep -vE ${excludes} | tr '\n' ' ')${adds}
+echo "Included files and directories: ${archives}"
+git archive $release_hash --prefix $tag/ -o $tarball ${archives}
```

Comment:
When I run it as is on my local, I get a fatalpathspec. However, when I substitute the variable of `$archives` with `$(git ls-tree --name-only -r HEAD | cut -d"/" -f1 | uniq | grep -vE ${excludes} | tr '\n' ' ')` directly, the tarball is correctly generated.

```
$ git archive $release_hash --prefix $tag/ -o $tarball ${archives}
fatal: pathspec '.asf.yaml .baseline .github .gitignore .travis.yml LICENSE NOTICE README.md api arrow baseline.gradle bundled-guava common core data deploy.gradle dev flink gradle.properties gradle gradlew hive-metastore hive-runtime jmh.gradle mr orc parquet pig settings.gradle spark-runtime spark spark2 spark3-runtime spark3 tasks.gradle version.txt versions.props ' did not match any files
```

Reply:
Also of note, is that the tarball is declared as `.tar.gz`, however, my output does not come gzipped.

`tarball=$tag.tar.gz`
https://github.com/apache/iceberg/blob/7da4aea3fb5a2d976305bbfbea23007fb385dacd/dev/source-release.sh#L65

Reply:
If we use the flag `--format .tar.gz` to `git archive`, I believe that should generate a truly gzipped tarball.

Reply:
Actualy ignore this comment related to gzipping, the output is correct and that flag just mucks things up. I'll delete these comments as to not clutter the PR.

Reply:
Ok. So I've I'm come up wth another solution that generates the proper tarball, akin to my second command there, but still broken up into chunks (vs doing WAY too much shell scripting and commands on a single line).

Instead of doing `git archive $release_hash --prefix $tag/ -o $tarball $(git ls-tree --name-only -r HEAD | cut -d"/" -f1 | uniq | grep -vE ${excludes} | tr '\n' ' ')`, where we inline what was previously the $archives variable, we can either write the $archives to a file and then cat that file out as part of the git archive or we can just echo the $archives variable in line (which seems to handle the weird string issues in my zsh shell, which is a relatively common shell).

If we wrote it to a file, that file wouldn't get included as it wouldn't be in the `git ls-tree`. We'd just want to `rm` it at the end to do some clean up.

So here's my proposed solution. It's a little more complex than using $archives inline, but my zsh shell seems to be interpreting it as a single string when we substitute it in and either of these approaches gets around that.

```bash
# Note I am removing the adds line as it is no longer needed because the `git ls-tree` will include .baseline
excludes="build|examples|jitpack.yml|python|site"  # excluded as they are not of use for releasing jars
# Alternativey, we might add the `.github` directory to the excludes list as it's not needed for releasing
echo "Excluded files and directories: ${excludes}"
# Generate the files to archive
# Note that we no longer need to append the $adds as the `.baseline` file comes from git ls-tree
archives=$(git ls-tree --name-only -r HEAD | cut -d"/" -f1 | uniq | grep -vE ${excludes} | tr '\n' ' ')
echo "Included files and directories: ${archives}"

# Two possible ways of achieving this.
# Way 1: cat the variable so as to make it no longer a string. This worked for me.
git archive $release_hash --prefix $tag/ -o $tarball $(cat $archives) 

# Way two: Write the archives to a file and then echo that file where we are currently placing $archives
echo $archives >> to_include.txt
git archive $release_hash --prefix $tag/ -o $tarball $(cat to_include.txt) # Also generates the tarball
rm to_include.txt
```

I'm still looking to see if there's a cleaner way to do this, but either of the above solutions will work across multiple shells. I think excluding zsh shell portability would be short sighted as it's now the default on the latest versions of macOS if I'm not mistaken.

Let me know what you think @rdblue. I'm personally fond of writing it out to a file.

Reply:
Oh, I also found that since we already created a tag and pushed it, we could possibly just use that tag on tag^{tree}. I've yet to find a way to make this work with our excludes list yet though.

Reply:
Scratch that, we can tag certain files or folders with `export-ignore` in the .gitattributes file. This is the proper way to exclude them from a git archive command as it's built into git itself. This would also allows us to skip the admittedly large amount of shell commands that are going into this. And while some might not be familiar with `.gitattributes`, I would suspect that a release engineer should hopefully be.

Here's the documentation on the git archive command: https://www.linux.org/docs/man1/git-archive.html
And here's a post discussing the usage of `export-ignore` in the `.gitattributes` file to properly ignore files. https://feeding.cloud.geek.nz/posts/excluding-files-from-git-archive/

I think this is the best approach as it uses the in built tools of git for generating this (so hopefully it should be better understood by release engineers), and it involves way less shell munging than currently.

An example .gitattributes file for the excludes we want would be something like
```
/build export-ignore
/examples export-ignore
.jitpack.yml export-ignore
/python export-ignore  # though do we really want to continue to exclude the python directory from the release?
/site export-ignore
```

The above complicated set of `excludes`, `archives, etc could instead be used off of the tag that is already pushed earlier in the script.

```bash
# Assuming the .gitattributes file is set up with the appropriate export-ignore
git archive --prefix $tag/ -o $tarball --worktree-attributes  $tag^{tree}
```

I am going to open a PR for adding a `.gitattributes` file so we can test with that. You can feel free to add it to this branch if you prefer, but I can't push to this branch so it's easier for me to create a separate PR.

Reply:
Thanks for looking into it more, @kbendick!